test/GatewayUpgradeFlow.js

```diff
@@ -0,0 +1,180 @@
+    });
+
+    it('should get the correct mint limiter address', async () => {
+        expect(await gateway.mintLimiter()).to.eq(mintLimiter.address);
+    });
+
+    it('should get the correct auth module', async () => {
+        expect(await gateway.authModule()).to.eq(auth.address);
+    });
+
+    it('auth module should have the correct owner', async () => {
+        expect(await auth.owner()).to.eq(gateway.address);
+    });
+
+    it('should get the correct token deployer', async () => {
+        expect(await gateway.tokenDeployer()).to.eq(tokenDeployer.address);
+    });
+
+    it('should schedule a proposal to upgrade AxelarGateway', async () => {
+        const commandID = 0;
+        const target = gateway.address;
+        const nativeValue = 0;
+        const timeDelay = 12 * 60 * 60;
+
+        const targetInterface = new Interface([
+            'function upgrade(address newImplementation, bytes32 newImplementationCodeHash, bytes calldata setupParams) external',
+        ]);
+        const newGatewayImplementation = await gatewayFactory.deploy(auth.address, tokenDeployer.address).then((d) => d.deployed());
+        const newGatewayImplementationCodeHash = await getBytecodeHash(newGatewayImplementation);
+        const setupParams = '0x';
+        const calldata = targetInterface.encodeFunctionData('upgrade', [
+            newGatewayImplementation.address,
+            newGatewayImplementationCodeHash,
+            setupParams,
+        ]);
+
+        const block = await ethers.provider.getBlock('latest');
+        const eta = block.timestamp + timeDelay;
+
+        const proposalHash = solidityKeccak256(['address', 'bytes', 'uint256'], [target, calldata, nativeValue]);
+
+        const payload = defaultAbiCoder.encode(
+            ['uint256', 'address', 'bytes', 'uint256', 'uint256'],
+            [commandID, target, calldata, nativeValue, eta],
+        );
+
+        await expect(interchainGovernance.executeProposalAction(governanceChain, governanceAddress, payload))
+            .to.emit(interchainGovernance, 'ProposalScheduled')
+            .withArgs(proposalHash, target, calldata, nativeValue, eta);
+    });
+
+    it('should execute an upgrade proposal on AxelarGateway', async () => {
+        const commandID = 0;
+        const target = gateway.address;
+        const nativeValue = 0;
+        const timeDelay = 12 * 60 * 60;
+
+        const targetInterface = new Interface([
+            'function upgrade(address newImplementation, bytes32 newImplementationCodeHash, bytes calldata setupParams) external',
+        ]);
+        const newGatewayImplementation = await gatewayFactory.deploy(auth.address, tokenDeployer.address).then((d) => d.deployed());
+        const newGatewayImplementationCodeHash = await getBytecodeHash(newGatewayImplementation);
+        const setupParams = '0x';
+        const calldata = targetInterface.encodeFunctionData('upgrade', [
```

Comment:
why not do target.encodeFunctionDate ...? instead of specifying the ABI explicitly

Reply:
I changed to define the interface using contract.interface.fragments